### PR TITLE
feat: allow to pass server-side flags to the client without cookies

### PR DESCRIPTION
**Pull request by masone**
We want to move away from storing application state in cookies that we intend to pass from the server-side to the client. While looking at https://autoricardo.atlassian.net/browse/CAR-3415 we found that this is possible through attaching the data to the express `req`, read it in `getInitialProps` of `_app` and wrap it with a `useMemo`. This way, the data is cached in `_app` during page transitions even when the client-side call to `getInitialProps` is missing the `req`

In this specific case, persisting feature flags in cookies cause issues for clients that - well... don't support cookies. We generate a new uuid for every client side transition, therefore generate too many MAU. We also run into problematic state when a user deletes that cookie, but keeps doing client side transitions

What looked like an easy rewrite, ended up escalating a bit. I did numerous breaking changes. Here's what changed:

- The express middleware that sets `req.ldData` from the project is moved to the pkg. This allows DH to migrate to server-side feature flags easily. One can use the pkg without the middleware, but that makes little sense
- The `req.ldData` has to be passed to the `LDProvider` where it's cached with a `useMemo` and therefore persisted across page transitions without cookies
- The project defines a `getLDUser` callback which returns the `LDUser` object. Listings will do this based on uuid, while DH will extract the logged in dealer from the token
- The feature flags from the server are directly passed through the provider to the consumers. All feature flags, both server and client side are readily available. We used to set state with the client and expose the feature flags from the client. This is not necessary at all. The project informs GA about the flags directly. There is no need to subscribe to events on the `LDClient`
- The `LDClient` still needs to be initialized in order to get pageview and click tracking in LD. The function `allFlags()` must be called or no events are sent to LD and experiments would not be possible
- All experiments in LD where we don't use experiments can be server-side and don't need to be exposed to the client. We are billed for the client-side flags, so this is great news
- Bot detection is now a concern of the pkg
- Removed HOC
- Renamed the `Provider`/`Consumer` to `LDProvider`/`LDConsumer`
- Changed the signature of both `LDProvider`/`LDConsumer`
- visitorId from query param is no longer stored to the cookies

### Benefits
- This will resolve misusing cookies for this case https://autoricardo.atlassian.net/browse/CAR-3415
- Server-side flags in DH should be easy now https://autoricardo.atlassian.net/browse/CAR-4426
- Should decrease metrics in LD that we're billed for

TODO
- [x] if (user.key && user.key !== "bot") - loosely tied logic with project. do based on empty user (or so)

### CODE REVIEW

File: src/context/index.tsx [hunk 1]
```diff
@@ -1,69 +1,45 @@
...
-const ProviderWithState: FC<ProviderProps> = ({
-  flags,
-  ldClientId,
-  ldUser,
-  children,
-}) => {
-  const [ldClient, setLdClient] = useState<LDContext>()
+const LDProvider: FC<Props> = ({ initialLDData, clientId, children }) => {
+  // persists the data initialized server-side on the client
+  const ldData = useMemo(() => initialLDData, [])
```
Comment (ufkaya) [inline src/context/index.tsx:23]: Nice usage of `useMemo` here but `initialLDData` is just an object and it is not optional so we should always get it when this provider is rendered, right? I think it is better to use it in the application, `useMemo(() => req.ldData, [])`, so we are more flexible and we can change the object if needed, wdyt?

File: src/requestHandler.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    if (!sdkKey) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        "No sdkKey passed to LDRequestHandler, will not load the feature flags form the launch darkly API"
+      )
+      return next()
+    }
+
+    if (
+      [/^\/_next/, /^\/static/, /\.\w+$/].find((matcher) => url.match(matcher))
```
Comment (lkappeler) [inline src/requestHandler.ts:73]: That looks pretty next specific, maybe name the handler getLDNextRequestHandler, btw didn't they change the defualt static to public so we would need to match that one as well...

File: src/requestHandler.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    return existingLDClient
+  }
+
+  const newLDClient = await createLDClient(sdkKey)
+  app.set("LDClient", newLDClient)
+  return newLDClient
+}
+
+const matchesBotUserAgent = (headers) => {
+  const botList = [
```
Comment (lkappeler) [inline src/requestHandler.ts:26]: Id still go with a 3rd party library for this

File: src/requestHandler.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    const isBot = matchesBotUserAgent(headers)
+    const user = getLDUser({ req, res, isBot })
+    const ldClient = await getLDClient(app, sdkKey)
+    const allFlags = await ldClient.allFlagsState(user)
+
+    const ldData: LDData = {
+      user,
+      allFlags: allFlags.toJSON(),
```
Comment (lkappeler) [inline src/requestHandler.ts:85]: Why are we converting the allFlags to a JSON object, dont we need to cenvert it back to an object if we wanna doe some handling in the next app with the flag data?

File: src/context/index.tsx (hunk 1 above)
Comment (masone) [reply src/context/index.tsx:23]: I think I get what you're hinting at and it's absolutely valid. `initialData` will be empty on page transitions as `getInitialProps` is called client-side and `req` is missing. The data is ignored on recurring renders and takes the memoized value. I agree that's a bit weird indeed.

I prefer to have this logic encapsulated in the pkg over imposing it on the projects though. On the initial render, the data could be changed in flight. But I believe it should not be changeable after that

File: src/requestHandler.ts (hunk 3 above)
Comment (masone) [reply src/requestHandler.ts:26]: yeah, we can do that in a separate PR. the experiment seems to be a success so far

File: src/requestHandler.ts (hunk 2 above)
Comment (masone) [reply src/requestHandler.ts:73]: it should not do much harm in a regular express application and the README says it's for next apps. I don't have. hard feelings here, let me know if you have :)

File: src/requestHandler.ts (hunk 4 above)
Comment (masone) [reply src/requestHandler.ts:85]: Good point. I vaguely remember that you can't assign unserialized objects to the request. The converting part must be implicit. Seems to have worked for a while now

File: src/requestHandler.ts (hunk 2 above)
Comment (masone) [inline src/requestHandler.ts:73]:
```suggestion
      [/^\/_next/, /^\/static/, /^\/public/, /\.\w+$/].find((matcher) => url.match(matcher))
```

Comment (masone) [reply src/requestHandler.ts:73]: this was solved in a nicer way before, but I did not want to impose the router dependency / knowledge on the pkg

File: src/context/index.tsx (hunk 1 above)
Comment (ufkaya) [reply src/context/index.tsx:23]: I see but TBH if we want to ensure to pass the right props, that part should be done outside of the component, but yeah..I guess it is just a preference

Comment (masone) [reply src/context/index.tsx:23]: I'll release it as-is and when we see an issue, I'll move it. Help me watching out for them! :)

**Pull request merged**